File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []

        def backTracking(s: str, openN: int, closeN: int):
            if len(s) == 2 * n:
                ans.append(s)
                return
            if openN < n:
                backTracking(s + '(', openN + 1, closeN)
            if closeN < openN:
                backTracking(s + ')', openN, closeN + 1)
        
        backTracking('', 0, 0)
        return ans

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    private ArrayList<String> ans;
    private int n;
    public List<String> generateParenthesis(int n) {
        this.ans = new ArrayList<>();
        this.n = n;
        backTracking("", 0, 0);
        return ans;
    }

    private void backTracking(String s, int leftN, int rightN) {
        if (s.length() == 2 * n) {
            ans.add(s);
            return;
        }
        if (leftN < n) {
            backTracking(s + "(", leftN + 1, rightN);
        }
        if (rightN < leftN) {
            backTracking(s + ")", leftN, rightN + 1);
        }
    }
}